_posts/2024-01-29-good-parallel-computer.md

```diff
@@ -0,0 +1,74 @@
+
+In the past, there were economic pressures towards replacing special-purpose circuitry with general purpose compute performance, but those incentives are shifting. Basically, if you're optimizing for number of transistors, then somewhat less efficient general purpose compute can be kept busy almost all the time, while special purpose hardware is only justified if there is high enough utilization in the workload. However, as Dennard scaling has ended and we're more constrained by power than transistor count, special purpose hardware starts winning more; it can simply be powered down if it isn't used by the workload. The days of a purely RISC computational model are probably over. What I'd *like* to see replacing it is an agile core (likely RISC-V) serving as the control function for a bunch of special-purpose accelerator extensions. That certainly is the model of the [Vortex] project among others.
+
+## Big grid of RISC-V
+
+There are many, many AI accellerators in the pipeline – see the [New Silicon for Supercomputers] talk for a great survey. One approach (definitely the one taken by the original [Google TPU]) is to sacrifice agility and make hardware that's specialized just for doing big matrix multiplications and essentially nothing else. Another approach, suitable for the low end, is a fairly vanilla VLIW microprocessor with big vector units, an architecture actually quite similar to existing DSPs. That is the approach taken by the [Qualcomm Hexagon]. Neither of these is suitable for running a workload like Vello.
+
+Far more interesting is the "big grid of RISC-V cores" approach. The idea is to achieve throughput simply by having many cores in parallel, a strategy generally in alignment with GPUs. However, all of the special purpose graphics hardware can be stripped out, leaving a focus on pure compute, so potentially a much simpler design. I'm aware of two, as both have provided substantial public detail, but I expect there are quite a few others.
+
+### Tenstorrent
+
+  + [TT-Metal] (Metalium)
+
+### Esperanto
+
+Another approach is [Esperanto], which is about 1000 efficiency RISC-V cores on a chip. The company way founded by Dave Ditzel, previously of Transmeta. The linked paper goes into a fair amount of detail and quantitative measurement. Not surprisingly, it focuses on the AI acceleration use case, but it also appears suitable for HPC workloads. Because it is at heard many CPUs, each running a program independently, it promises great agility. Unfortunately, there's no indication their software stack is open source, so it's hard for me to find out more.
+
```

Comment:
Typo: it is at heard

Reply:
Each running a portal independently => the same program? Same lifetime? IPC? How is this more agile?

Also, from the paper: "Four of these 8-core neighborhoods are put together along with 4 MB of memory to form a 32-core “Minion Shire.” The memory is implemented as four 1 MB SRAM banks connected to the neighborhoods through a 512-bit crossbar switch. These SRAM banks operate near the process-nominal supply voltage to allow higher density than the smaller caches within each core. Each bank can be partitioned by software to provide a mix of scratchpad memory, L2 cache private to the Shire, or L3 cache globally accessible across the entire chip via a global shared address space" => I have questions!! :-)